profiling/src/profile/mod.rs

```diff
@@ -1155,6 +1320,1224 @@ mod api_test {
+            },
+            api::ValueType {
+                r#type: "wall-time",
+                unit: "nanoseconds",
+            },
+            api::ValueType {
+                r#type: "cpu-time",
+                unit: "nanoseconds",
+            },
+        ];
+
+        let mut profile: Profile = Profile::builder().sample_types(sample_types).build();
+
+        // adding same offsets
+        let by_value_offsets: Vec<usize> = vec![0, 4];
+        let upscaling_info = UpscalingInfo::Poisson {
+            x: 1,
+            y: 100,
+            sampling_distance: 1,
+        };
+        profile
+            .add_upscaling_rules(
+                by_value_offsets.as_slice(),
+                "my label",
+                "coco",
+                upscaling_info,
+            )
+            .expect_err("Invalid offset");
+    }
+
+    #[test]
+    fn test_add_rule_with_offset_out_of_bound_poisson_function2() {
+        let sample_types = vec![
+            api::ValueType {
+                r#type: "samples",
+                unit: "count",
+            },
+            api::ValueType {
+                r#type: "wall-time",
+                unit: "nanoseconds",
+            },
+            api::ValueType {
+                r#type: "cpu-time",
+                unit: "nanoseconds",
+            },
+        ];
+
+        let mut profile: Profile = Profile::builder().sample_types(sample_types).build();
+
+        // adding same offsets
+        let by_value_offsets: Vec<usize> = vec![0, 4];
+        let upscaling_info = UpscalingInfo::Poisson {
+            x: 100,
+            y: 1,
+            sampling_distance: 1,
+        };
+        profile
+            .add_upscaling_rules(
+                by_value_offsets.as_slice(),
+                "my label",
+                "coco",
+                upscaling_info,
+            )
+            .expect_err("Invalid offset");
+    }
+
+    #[test]
+    fn test_add_rule_with_offset_out_of_bound_poisson_function3() {
+        let sample_types = vec![
+            api::ValueType {
+                r#type: "samples",
+                unit: "count",
+            },
+            api::ValueType {
+                r#type: "wall-time",
+                unit: "nanoseconds",
+            },
+            api::ValueType {
+                r#type: "cpu-time",
+                unit: "nanoseconds",
+            },
+        ];
+
+        let mut profile: Profile = Profile::builder().sample_types(sample_types).build();
+
+        // adding same offsets
+        let by_value_offsets: Vec<usize> = vec![0, 4];
+        let upscaling_info = UpscalingInfo::Poisson {
+            x: 1100,
+            y: 100,
+            sampling_distance: 1,
+        };
+        profile
+            .add_upscaling_rules(
+                by_value_offsets.as_slice(),
+                "my label",
+                "coco",
+                upscaling_info,
+            )
+            .expect_err("Invalid offset");
+    }
```

Comment:
Minor: Since we already test both x and y out of bounds, I'm not sure testing "both out of bounds" adds a lot, it exercises the same code as above. (Unless I'm missing something?)

Reply:
I agree that feels redundant, I wonder what happen if the code change and this test is removed and nothing fails ?
because in terms of tests cases, this is a valid one.

Reply:
Maybe... my thinking was -- what weird change would need to happen for the "x out of bounds" and "y out of bounds" individual tests to still pass, and the "both x and y out of bounds" test to fail?

Nevertheless, indeed the reason I marked my feedback as minor is that it's a small thing, so if you're not convinced, I can definitely live with that ;)